### DISCUSSION TITLE
Building and running "from scratch" on a non-Arch distro

### DISCUSSION THREAD

**Discussion by mato** [nwg-piotr/nwg-shell#34] [General]
Hi Piotr,

this is related to #18. I'd like to try and get nwg-shell working on Debian with sway, and once that works, try to improve it to work with the labwc stacking compositor, since coming from Openbox I much prefer that to sway.

I'm fairly familiar with Go, but haven't used Python much recently and am missing a "recipe" to get the Python components working, starting with nwg-panel. In order to understand how things fit together I don't want to run anything via an installer, or indeed install any components system-wide yet. So, I'm looking for a way to run the components "in place" for development.

For nwg-panel, so far I have the following:

1. Install the following Debian packages: `python3 python3-pip python3-psutil python3-cairo libgtk-layer-shell-dev`.
2. From a clean source tree, run: `pip install -e .`
3. Run `$HOME/.local/bin/nwg-panel`.

(To be clear, this is on a machine running Debian testing/bookworm. I'm not even going to try with stable, as the various Wayland components are too old.)

However, this fails with the exception:

```
Traceback (most recent call last):
  File "/home/mato/.local/bin/nwg-panel", line 33, in <module>
    sys.exit(load_entry_point('nwg-panel', 'gui_scripts', 'nwg-panel')())
<... snip ...>
  File "/home/mato/src/nwg-shell/nwg-panel/nwg_panel/main.py", line 44, in <module>
    from nwg_panel.modules.playerctl import Playerctl
  File "/home/mato/src/nwg-shell/nwg-panel/nwg_panel/modules/playerctl.py", line 8, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

I can't figure out why it can't load the `Playerctl` module. What am I doing wrong?

**Comment by nwg-piotr** [comment]
Hi @mato,

I'll be glad to help, for now except the labwc thing. The shell will remain sway-centered until we have a true and real team of developers.

The playerctl module crashes for you while trying to import the `python-requests` module (or whatever it's called in Debian). It's used to get the album cover, when given as a remote url (e.g. Epiphany browser does so). See [here](https://github.com/nwg-piotr/nwg-panel/blob/518ac0a2beeaea293c586125a6da7c9bb3bfac2d/nwg_panel/modules/playerctl.py#L74).

BTW: dunno what `pip install -e` means. On Arch using pip is always a bad idea, and I use setup.py on my development machine. See [sample install script](https://github.com/nwg-piotr/nwg-panel/blob/master/install.sh).

**Comment by mato** [comment]
Hi Piotr,

> I'll be glad to help, for now except the labwc thing. The shell will remain sway-centered until we have a true and real team of developers.

Sure. My first goal is to get it running at all, in a reproducible way. I can use sway, just prefer very much not to, having been using Openbox for the last 15 years or so.

> The playerctl module crashes for you while trying to import the python-requests module.

Got it, thanks. Installing `python3-requests` fixes it, and nwg-panel runs. So far so good.

It now complains about `python-dasbus` not being installed, but installing the Debian `python3-dasbus` package doesn't seem to change that, I'll look into it.

> BTW: dunno what pip install -e means. On Arch using pip is always a bad idea, and I use setup.py ...

I used to use setup.py years ago when working on Python 2 code. When I tried that now, I got a ton of exceptions and deprecation warnings sternly telling me not to, so I went looking and found this: https://packaging.python.org/en/latest/tutorials/installing-packages/#installing-from-a-local-src-tree.

The `-e` option just says "install the package in editable (development) mode". I presume that will let me edit the code and re-run without needing to run `pip install` again.

It seems that `pip` is the way to do things nowadays, and indeed it installed the `pycairo` dependency for me locally in `$HOME` since you list it in `requirements.txt`. Perhaps we should add all the other dependencies in there, it would make it easier for folks starting from scratch?

**Comment by nwg-piotr** [comment]
Your choice, dependencies and installation method remain on packager's side. From the code point of view, it doesn't matter which way you make the python module available. E.g. I don't use packages at all on my development machine, but only while testing.

**Comment by kode54** [comment]
labwc will outright fail to even run anything in this project which currently depends on Sway/i3 IPC socket functionality. This includes nwg-shell-config when it hits the output enumeration, and also the process list tool.

The most that will work is things using protocols such as foreign toplevel.

(I don't really want to have to use Sway, because I hate the very organization of having to throw all my windows into a tiled grid.)

**Comment by nwg-piotr** [comment]
On the 'split' branch there are 2 different nwg-shell-config versions: aimed at sway, depending on i3ipc, and aimed at Hyprland, reading from their socket and socket2. Labwc would require a 3rd config app. Also the dock would need to be written in a 3rd version. For now we have nwg-dock and nwg-dock-hyprland. The rest should work well, apart from the fact, that nwg-panel has no labwc-specific modules.

**Reply by kode54** [reply]
Ah, I wasn't aware there were other branches for the other desktops. I'll see if I can help out with this.

**Reply by nwg-piotr** [reply]
Hyprland is still work in progress. And I have no plan to support anything but sway and Hyprland, to be honest. Much too much work for a single hobbyist. I'm open to reasonable contributions, however.
